pkg/hbone/server.go

```diff
@@ -16,17 +16,31 @@ package hbone
 
 import (
 	"context"
+	"crypto/tls"
 	"net"
 	"net/http"
 	"sync"
 	"time"
 
+	"github.com/cbeuw/connutil"
 	"golang.org/x/net/http2"
 
 	"istio.io/istio/pkg/h2c"
 )
 
+func NewDoubleHBONEServer(tlsConfig *tls.Config) *http.Server {
```

Comment:
IMO a pair of servers ( outer and inner ) would more accurately test the protocol. It is not expected for a native app to handle 2x H2 - just on the client side. 

Having a native go 'outer' H2 router would also help as a standalone minimal E/W gateway.

Reply:
Yeah I considered that; it was actually the first implementation that I got working. My reasoning for doing it all in memory is more test speed and not requiring an extra port.

Reply:
Oh that's basically what this code does; the inner server is the exact same single HBONE server that already existed: https://github.com/istio/istio/pull/54285/files/5f848e2d9a4ee962c3ade89c053e04f26ed71e89#diff-3dc57dc389b89e1156ea75bebb163573d4e5513460c434b4b6bf7e4b2917211bR89.

Reply:
My point was that it would be useful to split/isolate this - so the 'outer' side can be used with a regular Dialer or http client. Maybe I'm missing something and that is already available ?

Reply:
Close - I'm looking for this kind of test, but with client or server substituted with a non-Istio component that implements double H2 standard.

Reply:
So to clarify, you'd want to see something like firefox, nginx, etc. implement double h2 proxying and use that as a test/

Reply:
@costinm just to be clear: this implementation is just a go in memory double HBONE implementation written from first principles. No users will use either of these; the ztunnel and east west gateway implementations don't use this at all. Doesn't that serve as a sort of check?